metaflow/plugins/airflow/airflow_utils.py

```diff
@@ -0,0 +1,477 @@
+from collections import defaultdict
+import os
+import typing
+import json
+import time
+import random
+import hashlib
+import re
+from datetime import timedelta, datetime
+
+
+class KubernetesProviderNotFound(Exception):
+    headline = "Kubernetes provider not found"
+
+
+LABEL_VALUE_REGEX = re.compile(r"^[a-zA-Z0-9]([a-zA-Z0-9\-\_\.]{0,61}[a-zA-Z0-9])?$")
+
+TASK_ID_XCOM_KEY = "metaflow_task_id"
+
+# AIRFLOW_TASK_ID_TEMPLATE_VALUE will work for linear/branched workflows.
+# ti.task_id is the stepname in metaflow code.
+# AIRFLOW_TASK_ID_TEMPLATE_VALUE uses a jinja filter called `task_id_creator` which helps
+# concatenate the string using a `/`. Since run-id will keep changing and stepname will be
+# the same task id will change. Since airflow doesn't encourage dynamic rewriting of dags
+# we can rename steps in a foreach with indexes (eg. `stepname-$index`) to create those steps.
+# Hence : Foreachs will require some special form of plumbing.
+# https://stackoverflow.com/questions/62962386/can-an-airflow-task-dynamically-generate-a-dag-at-runtime
+AIRFLOW_TASK_ID_TEMPLATE_VALUE = "arf-{{ [run_id, ti.task_id ] | task_id_creator  }}"
+
+
+def sanitize_label_value(val):
```

Comment:
JFYI - I will be removing this functionality in @kubernetes. Hashing this way has limited utility right now and it's better for us to understand the use cases rather than shipping a breaking change to the user later.